### PR TITLE
Hive 23301

### PR DESCRIPTION
**Pull request by simhadri-g**
UDF to map bits to privileges.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -38,3 +38,4 @@ standalone-metastore/metastore-server/src/gen/version
 launch.json
 settings.json
 kafka-handler/src/test/gen
+
```
Comment (aasha) [inline .gitignore:41]: unused can be removed

Comment (aasha) [reply .gitignore:41]: Is the empty udf_split_map_privs.q.out file needed?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java [hunk 2]
```diff
@@ -285,6 +285,7 @@
     system.registerGenericUDF("quote", GenericUDFQuote.class);
     system.registerGenericUDF("nvl", GenericUDFCoalesce.class); //HIVE-20961
     system.registerGenericUDF("split", GenericUDFSplit.class);
+    system.registerGenericUDF("split_map_privs", UDFSplitMapPrivs.class);
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java:288]: Rename to GenericUDFStringToPrivilege

File: .gitignore (hunk 1 above)
Comment (simhadri-g) [reply .gitignore:41]: > Is the empty udf_split_map_privs.q.out file needed?
no, it is generated while building

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java (hunk 2 above)
Comment (simhadri-g) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java:288]: ok

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+ *     output: "  ["UPDATE", "CREATE", "ALTER", "INDEX"]"
+ */
+public class GenericUDFStringToPrivilege extends GenericUDF {
+  private transient ObjectInspectorConverters.Converter[] converters;
+  private transient Pattern constPattern;
+
+  private PrivilegeMap privsMap = new PrivilegeMap();
+
+  @Override public ObjectInspector initialize(ObjectInspector[] arguments) throws UDFArgumentException {
+    if (arguments.length != 2) {
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java:78]: You can refer public class UDFYear extends GenericUDF { and use the same validation. like checkArgsSize(arguments, 1, 1);
    checkArgPrimitive(arguments, 0);

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      constPattern = Pattern.compile(((ConstantObjectInspector) rightArg).
+          getWritableConstantValue().toString());
+    }
+
+    return ObjectInspectorFactory
+        .getStandardListObjectInspector(PrimitiveObjectInspectorFactory.writableStringObjectInspector);
+  }
+
+  @Override public Object evaluate(DeferredObject[] arguments) throws HiveException {
+    assert (arguments.length == 2);
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java:99]: if its a constant ..then user may not give the second arguemnt

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    if (arguments[0].get() == null || arguments[1].get() == null) {
+      return null;
+    }
+
+    Text s = (Text) converters[0].convert(arguments[0].get());
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java:105]: The converter is based on user argument to initialize ..and here the conversion is always to text ..that may cause issue

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    privilegeMap.put(9, "ALL");
+
+    return privilegeMap;
+  }
+}
+
+/**
+ * UDFSplitMapPrivs.
+ * "_FUNC_(str, regex) - Splits binary str and maps to privilege type "
+ *      "Example: > SELECT _FUNC_('0 1 1 0 1 1 0 0 0', ' ') FROM src LIMIT 1;"
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java:68]: i think ..make it simple ..no need to take any converter ...just accept string and split is using space ..

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    Text s = (Text) converters[0].convert(arguments[0].get());
+    ArrayList<Text> result = new ArrayList<Text>();
+    int index = 0;
+    Map<Integer, String> privs = privsMap.getPrivilegeMap();
+
+    if (constPattern == null) {
+      Text regex = (Text) converters[1].convert(arguments[1].get());
+      for (String str : s.toString().split(regex.toString(), -1)) {
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java:112]: if else can be combined to one ..after initializing the pattern before loop

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        if ("1".equals(str)) {
+          result.add(new Text(privs.get(index)));
+        }
+        index++;
+      }
+    } else {
+      for (String str : constPattern.split(s.toString(), -1)) {
+        if ("1".equals(str)) {
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToPrivilege.java:120]: should check for validity like getting more that 10 splits

File: itests/src/test/resources/testconfiguration.properties [hunk 9]
```diff
@@ -1698,6 +1698,7 @@ minillaplocal.query.files=\
   smblimit.q,\
   specialChar.q,\
   split.q,\
+  split_map_privs.q,\
```
Comment (maheshk114) [inline itests/src/test/resources/testconfiguration.properties:1701]: GenericUDFSplit can be re-used ..i think this udf is not required

**Comment by belugabehr** [comment]
Marked as resolved in JIRA.  Please reopen if required.

https://issues.apache.org/jira/browse/HIVE-23301

**Pull request closed without merging**